File: src/test/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/eksportering/autoeksport/EksporteringPerStatistikkKategoriServiceMockTest.java
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport;

import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.__2019_3;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.__2019_4;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.__2020_1;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.__2020_2;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.landSykefravær;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.sykefraværsstatistikkLandSiste4Kvartaler;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.sykefraværsstatistikkVirksomhet;
import static no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport.EksporteringServiceTestUtils.virksomhetSykefraværMedKategori;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.Arrays;
import java.util.List;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.EksporteringRepository;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.SykefraværsstatistikkTilEksporteringRepository;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.VirksomhetEksportPerKvartal;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.Orgnr;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.felles.ÅrstallOgKvartal;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.importering.SykefraværsstatistikkVirksomhetUtenVarighet;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.kafka.KafkaService;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.Statistikkategori;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefravar.SykefraværMedKategori;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.UmaskertSykefraværForEttKvartal;
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.summert.SykefraværRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.ArgumentCaptor;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
public class EksporteringPerStatistikkKategoriServiceMockTest {

  @Mock
  private SykefraværsstatistikkTilEksporteringRepository sykefraværsstatistikkTilEksporteringRepository;

  @Mock
  private SykefraværRepository sykefraværRepository;
  @Mock
  private EksporteringRepository eksporteringRepository;

  @Mock
  private KafkaService kafkaService;

  private EksporteringPerStatistikkKategoriService service;

  @Captor
  ArgumentCaptor<ÅrstallOgKvartal> årstallOgKvartalArgumentCaptor;
  @Captor
  ArgumentCaptor<Statistikkategori> statistikkategoriArgumentCaptor;
  @Captor
  ArgumentCaptor<String> identifikatorArgumentCaptor;
  @Captor
  ArgumentCaptor<SykefraværMedKategori> sykefraværMedKategoriArgumentCaptor;

  @Captor
  ArgumentCaptor<SykefraværFlereKvartalerForEksport> sykefraværFlereKvartalerForEksportArgumentCaptor;

  @BeforeEach
  public void setUp() {
    service = new EksporteringPerStatistikkKategoriService(sykefraværRepository,
        sykefraværsstatistikkTilEksporteringRepository, eksporteringRepository, kafkaService, true);
  }

  @Test
  public void eksporterSykefraværsstatistikkLand__sender_riktig_melding_til_kafka_og_returnerer_antall_meldinger_sendt() {
    List<UmaskertSykefraværForEttKvartal> umaskertSykefraværForEttKvartalListe = sykefraværsstatistikkLandSiste4Kvartaler(
        __2020_2);
    when(sykefraværRepository.hentUmaskertSykefraværForNorge(any())).thenReturn(
        umaskertSykefraværForEttKvartalListe);
    when(kafkaService.sendTilStatistikkKategoriTopic(any(), any(), any(), any(), any())).thenReturn(
        true);

    int antallEksporterte = service.eksporterSykefraværsstatistikkLand(__2020_2);

    verify(kafkaService).sendTilStatistikkKategoriTopic(årstallOgKvartalArgumentCaptor.capture(),
        statistikkategoriArgumentCaptor.capture(), identifikatorArgumentCaptor.capture(),
        sykefraværMedKategoriArgumentCaptor.capture(),
        sykefraværFlereKvartalerForEksportArgumentCaptor.capture());

    assertThat(årstallOgKvartalArgumentCaptor.getValue()).isEqualTo(__2020_2);
    assertThat(statistikkategoriArgumentCaptor.getValue()).isEqualTo(Statistikkategori.LAND);
    assertThat(identifikatorArgumentCaptor.getValue()).isEqualTo("NO");
    EksporteringServiceTestUtils.assertEqualsSykefraværMedKategori(landSykefravær,
        sykefraværMedKategoriArgumentCaptor.getValue());
    EksporteringServiceTestUtils.assertEqualsSykefraværFlereKvartalerForEksport(
        new SykefraværFlereKvartalerForEksport(umaskertSykefraværForEttKvartalListe),
        sykefraværFlereKvartalerForEksportArgumentCaptor.getValue());
    assertThat(antallEksporterte).isEqualTo(1);
  }

  @Test
  public void eksporterSykefraværsstatistikkVirksomhet__sender_riktig_melding_til_kafka_og_returnerer_antall_meldinger_sendt() {
    // 1- Mock det database og repositories returnerer. Samme med KafkaService
    List<SykefraværsstatistikkVirksomhetUtenVarighet> allData = List.of(
        sykefraværsstatistikkVirksomhet(__2020_2, "987654321"),
        sykefraværsstatistikkVirksomhet(__2020_1, "987654321"),
        sykefraværsstatistikkVirksomhet(__2019_4, "987654321"),
        sykefraværsstatistikkVirksomhet(__2019_3, "987654321"));
    when(sykefraværsstatistikkTilEksporteringRepository.hentSykefraværAlleVirksomheter(__2019_3,
        __2020_2)).thenReturn(allData);
    when(kafkaService.sendTilStatistikkKategoriTopic(any(), any(), any(), any(), any())).thenReturn(
        true);
    when(eksporteringRepository.hentVirksomhetEksportPerKvartal(__2020_2)).thenReturn(Arrays.asList(
        new VirksomhetEksportPerKvartal(new Orgnr("987654321"), new ÅrstallOgKvartal(2022, 2),
            false)));

    // 2- Kall tjenesten
    int antallEksporterte = service.eksporterSykefraværsstatistikkVirksomhet(__2020_2,
        EksporteringBegrensning.build().utenBegrensning());

    // 3- Sjekk hva Kafka har fått
    verify(kafkaService).sendTilStatistikkKategoriTopic(årstallOgKvartalArgumentCaptor.capture(),
        statistikkategoriArgumentCaptor.capture(), identifikatorArgumentCaptor.capture(),
        sykefraværMedKategoriArgumentCaptor.capture(),
        sykefraværFlereKvartalerForEksportArgumentCaptor.capture());

    assertThat(årstallOgKvartalArgumentCaptor.getValue()).isEqualTo(__2020_2);
    assertThat(statistikkategoriArgumentCaptor.getValue()).isEqualTo(Statistikkategori.VIRKSOMHET);
    assertThat(identifikatorArgumentCaptor.getValue()).isEqualTo("987654321");

    EksporteringServiceTestUtils.assertEqualsSykefraværMedKategori(virksomhetSykefraværMedKategori,
        sykefraværMedKategoriArgumentCaptor.getValue());
    assertThat(antallEksporterte).isEqualTo(1);
  }

  @Test
  public void eksporterSykefraværsstatistikkVirksomhet__returnerer_korrekt_antall_meldinger_sendt() {
    List<SykefraværsstatistikkVirksomhetUtenVarighet> allData = List.of(
        sykefraværsstatistikkVirksomhet(__2020_2, "987654321"),
        sykefraværsstatistikkVirksomhet(__2020_1, "987654321"),
        sykefraværsstatistikkVirksomhet(__2019_4, "987654321"),
        sykefraværsstatistikkVirksomhet(__2019_3, "987654321"),
        sykefraværsstatistikkVirksomhet(__2020_2, "987654322"),
        sykefraværsstatistikkVirksomhet(__2020_1, "987654322"),
        sykefraværsstatistikkVirksomhet(__2019_4, "987654322"));

    when(sykefraværsstatistikkTilEksporteringRepository.hentSykefraværAlleVirksomheter(__2019_3,
        __2020_2)).thenReturn(allData);
    when(kafkaService.sendTilStatistikkKategoriTopic(any(), any(), any(), any(), any())).thenReturn(
        true);
    when(eksporteringRepository.hentVirksomhetEksportPerKvartal(__2020_2)).thenReturn(Arrays.asList(
        new VirksomhetEksportPerKvartal(new Orgnr("987654321"), new ÅrstallOgKvartal(2022, 2),
            false),
        new VirksomhetEksportPerKvartal(new Orgnr("987654322"), new ÅrstallOgKvartal(2022, 2),
            false)));

    int antallEksporterte = service.eksporterSykefraværsstatistikkVirksomhet(__2020_2,
        EksporteringBegrensning.build().utenBegrensning());

    assertThat(antallEksporterte).isEqualTo(2);
  }

  @Test
  public void eksporterSykefraværsstatistikkVirksomhet__eksporterer_til_og_med_bedrifter_uten_statistikk() {
    List<SykefraværsstatistikkVirksomhetUtenVarighet> allData = List.of(
        sykefraværsstatistikkVirksomhet(__2020_2, "987654321"),
        sykefraværsstatistikkVirksomhet(__2020_1, "987654321"));

    when(sykefraværsstatistikkTilEksporteringRepository.hentSykefraværAlleVirksomheter(__2019_3,
        __2020_2)).thenReturn(allData);
    when(kafkaService.sendTilStatistikkKategoriTopic(any(), any(), any(),
        any(SykefraværMedKategori.class),
        any(SykefraværFlereKvartalerForEksport.class))).thenReturn(true);
    when(eksporteringRepository.hentVirksomhetEksportPerKvartal(__2020_2)).thenReturn(Arrays.asList(
        new VirksomhetEksportPerKvartal(new Orgnr("987654321"), new ÅrstallOgKvartal(2022, 2),
            false),
        new VirksomhetEksportPerKvartal(new Orgnr("987654322"), new ÅrstallOgKvartal(2022, 2),
            false)));

    int antallEksporterte = service.eksporterSykefraværsstatistikkVirksomhet(__2020_2,
        EksporteringBegrensning.build().utenBegrensning());

    assertThat(antallEksporterte).isEqualTo(2);
  }
}

File: src/test/java/no/nav/arbeidsgiver/sykefravarsstatistikk/api/eksportering/autoeksport/EksporteringPerStatistikkKategoriServiceMockTest.kt
package no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.autoeksport

import net.javacrumbs.jsonunit.assertj.assertThatJson
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.eksportering.SykefraværsstatistikkTilEksporteringRepository
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.infrastruktur.config.KafkaTopic
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.kafka.KafkaService
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.integrasjoner.kafka.dto.StatistikkategoriKafkamelding
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.Statistikkategori
import no.nav.arbeidsgiver.sykefravarsstatistikk.api.statistikk.sykefraværshistorikk.summert.SykefraværRepository
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.kotlin.*

class EksporteringPerStatistikkKategoriServiceMockTest {
    private val sykefraværRepository = mock<SykefraværRepository>()
    private val sykefraværsstatistikkTilEksporteringRepository = mock<SykefraværsstatistikkTilEksporteringRepository>()
    private val kafkaService = mock<KafkaService>()
    private var service: EksporteringPerStatistikkKategoriService = EksporteringPerStatistikkKategoriService(
        sykefraværRepository,
        sykefraværsstatistikkTilEksporteringRepository,
        kafkaService,
        true,
    )

    private val statistikkategoriKafkameldingCaptor = argumentCaptor<StatistikkategoriKafkamelding>()

    @BeforeEach
    fun setUp() {
        service = EksporteringPerStatistikkKategoriService(
            sykefraværRepository,
            sykefraværsstatistikkTilEksporteringRepository,
            kafkaService,
            true
        )
    }

    @Test
    fun eksporterSykefraværsstatistikkLand__sender_riktig_melding_til_kafka() {
        val umaskertSykefraværForEttKvartalListe =
            EksporteringServiceTestUtils.sykefraværsstatistikkLandSiste4Kvartaler(EksporteringServiceTestUtils.__2020_2)
        whenever(sykefraværRepository.hentUmaskertSykefraværForNorge(any()))
            .thenReturn(umaskertSykefraværForEttKvartalListe)

        service.eksporterSykefraværsstatistikkLand(EksporteringServiceTestUtils.__2020_2)

        verify(kafkaService)
            .sendMelding(
                statistikkategoriKafkameldingCaptor.capture(),
                eq(KafkaTopic.SYKEFRAVARSSTATISTIKK_LAND_V1)
            )

        assertThatJson(statistikkategoriKafkameldingCaptor.firstValue.innhold) {
            isObject
            node("kategori").isString.isEqualTo(Statistikkategori.LAND.name)
            node("kode").isString.isEqualTo("NO")
            node("sistePubliserteKvartal").isObject.node("årstall").isString.isEqualTo("2020")
        }
    }

    @Test
    fun eksporterSykefraværsstatistikkVirksomhet__sender_riktig_melding_til_kafka() {
        // 1- Mock det database og repositories returnerer. Samme med KafkaService
        val allData = listOf(
            EksporteringServiceTestUtils.sykefraværsstatistikkVirksomhet(
                EksporteringServiceTestUtils.__2020_2,
                "987654321"
            ),
            EksporteringServiceTestUtils.sykefraværsstatistikkVirksomhet(
                EksporteringServiceTestUtils.__2020_1,
                "987654321"
            ),
            EksporteringServiceTestUtils.sykefraværsstatistikkVirksomhet(
                EksporteringServiceTestUtils.__2019_4,
                "987654321"
            ),
            EksporteringServiceTestUtils.sykefraværsstatistikkVirksomhet(
                EksporteringServiceTestUtils.__2019_3,
                "987654321"
            )
        )
        whenever(
            sykefraværsstatistikkTilEksporteringRepository.hentSykefraværAlleVirksomheter(
                EksporteringServiceTestUtils.__2019_3, EksporteringServiceTestUtils.__2020_2
            )
        )
            .thenReturn(allData)

        // 2- Kall tjenesten
        service.eksporterSykefraværsstatistikkVirksomhet(
            EksporteringServiceTestUtils.__2020_2
        )

        // 3- Sjekk hva Kafka har fått
        verify(kafkaService)
            .sendMelding(
                statistikkategoriKafkameldingCaptor.capture(),
                eq(KafkaTopic.SYKEFRAVARSSTATISTIKK_VIRKSOMHET_V1)
            )

        assertThatJson(statistikkategoriKafkameldingCaptor.firstValue.innhold) {
            isObject
            node("kategori").isString.isEqualTo(Statistikkategori.VIRKSOMHET.name)
            node("kode").isString.isEqualTo("987654321")
            node("sistePubliserteKvartal").isObject.node("årstall").isString.isEqualTo("2020")
            node("siste4Kvartal").isObject.node("prosent").isString.isEqualTo("10000000000")
        }

    }


    @Test
    fun eksporterSykefraværsstatistikkNæring__sender_riktig_melding_til_kafka() {
        // 1- Mock det database og repositories returnerer. Samme med KafkaService
        val allData = listOf(
            EksporteringServiceTestUtils.sykefraværsstatistikkNæring(EksporteringServiceTestUtils.__2020_2, "11"),
            EksporteringServiceTestUtils.sykefraværsstatistikkNæring(EksporteringServiceTestUtils.__2020_1, "11"),
            EksporteringServiceTestUtils.sykefraværsstatistikkNæring(EksporteringServiceTestUtils.__2019_4, "11"),
            EksporteringServiceTestUtils.sykefraværsstatistikkNæring(EksporteringServiceTestUtils.__2019_3, "11")
        )

        whenever(
            sykefraværsstatistikkTilEksporteringRepository.hentSykefraværAlleNæringerFraOgMed(any())
        )
            .thenReturn(allData)

        // 2- Kall tjenesten
        service.eksporterSykefraværsstatistikkNæring(EksporteringServiceTestUtils.__2020_2)

        // 3- Sjekk hva Kafka har fått
        verify(kafkaService)
            .sendMelding(
                statistikkategoriKafkameldingCaptor.capture(),
                eq(KafkaTopic.SYKEFRAVARSSTATISTIKK_NARING_V1)
            )

        assertThatJson(statistikkategoriKafkameldingCaptor.firstValue.innhold) {
            isObject
            node("kategori").isString.isEqualTo(Statistikkategori.NÆRING)
            node("kode").isString.isEqualTo("11")
            node("sistePubliserteKvartal").isObject.node("årstall").isEqualTo("2020").node("kvartal").isString.isEqualTo("2")
        }

        assertThatJson(statistikkategoriKafkameldingCaptor.firstValue.nøkkel) {
            isObject
            node("kategori").isString.isEqualTo(Statistikkategori.NÆRING)
            node("kode").isString.isEqualTo("11")
            node("kvartal").isString.isEqualTo("2")
            node("årstall").isString.isEqualTo("2020")
        }
    }
}